### ISSUE TITLE
[feature] Let me see the federation status of a post I've made

### ISSUE THREAD

**Issue by hikari-no-yume** [superseriousbusiness/gotosocial#2295]
### Is your feature request related to a problem ?

I've been using GtS to host a personal instance (just for me) for the past year or so. A frequent frustration is that I'll make a post and, apparently, nobody else has seen it, and I don't know why. I often will go to my alt account on a certain well-known Mastodon instance in order to see if it at least arrived on _that_ other fediverse instance.

I can only assume that this is happening because it just takes a long time to notify the hundreds of different instances that have users that follow me. Or, it could be because something's going wrong when notifying those instances. The point is, it's frustrating having no idea what's going on.

### Describe the solution you'd like.

I would like there to be an option in the admin area that lets me see the federation status for a post from my instance. What I mean by that is: what instances have been notified, which haven't yet, and which have encountered problems?

Alternatively, I would like there to be an option in the admin area to see the backlog of outgoing notifications (or whatever they're called).

### Describe alternatives you've considered.

I often try to use `tail -f /var/log/gotosocial/gotosocial.log`, but to be honest there is so much activity and so little information for each item, so it's very hard for me to interpret it.

**Comment by hikari-no-yume** [comment]
If it's helpful: I personally suspect that there might be some broken/dead instances that my instance federates with, which could be blocking the queue for the actually live instances, when federating a new post. Is that possible? I don't know how I'd check, which is kind of why I want this feature.

**Comment by tsmethurst** [comment]
This is quite a large request! We do intend when time allows to implement some functionality where you can see failed message deliveries in the admin panel, so that will help diagnose federation issues, but per-post federation status would probably cause quite a bit of ballooning in terms of database storage size. Not saying it's impossible, but we'd have to think carefully about it -- so it won't be a quick fix/feature to implement.

>  I personally suspect that there might be some broken/dead instances that my instance federates with, which could be blocking the queue for the actually live instances

GoToSocial uses parallel queues for message delivery, described here under `advanced-sender-multiplier`: https://docs.gotosocial.org/en/latest/configuration/advanced/. When a message fails to be delivered because of errors on the receiving end, the sender that picked up that particular delivery job will retry for about 2 minutes. If you're trying to send to multiple instances with issues at once, it's possible that all your senders will get *temporarily* clogged up, so the message may take a while to deliver overall.

If you feel that federation is laggy, you can try tuning this by increasing your `advanced-sender-multiplier` a little bit. If you set it to a very high number, there's essentially no limit on the amount of outgoing send attempts that will be processed at once, however this will cause a spike of high CPU usage when sending messages from an account with lots of followers.

**Comment by hikari-no-yume** [comment]
> per-post federation status would probably cause quite a bit of ballooning in terms of database storage size

Ah, I understand. Well, I'd be happy with just being able to see recent failed deliveries, and ideally also upcoming deliveries, without them having to be specific to a particular post or stored in the DB. What I want is just to be able to get a bit more insight into what's going on.

**Comment by tsmethurst** [comment]
> What I want is just to be able to get a bit more insight into what's going on.

Yeah same! Federation being a "black box" is rather annoying. We'll work on adding something when time permits!

**Comment by hikari-no-yume** [comment]
By the way, about this sort of clogged pipes hypothesis, does GtS do anything to track instances that repeatedly have delivery issues? I wonder if it could improve average delivery times by sending persistently problematic instances to the back of the queue or something.

**Comment by tsmethurst** [comment]
No not yet, that's a good idea though. I think at some point we'll also offer some kind of facility where if an instance is MIA for like 7 days or something, you'll get a suggestion in the admin panel to consider defederating from that instance so that it stops causing problems.

**Comment by NyaaaWhatsUpDoc** [comment]
Honestly this feels like the kind of thing you could probably figure out with tracing enabled if we hook it into the http client :thinking:

**Comment by tsmethurst** [comment]
> Honestly this feels like the kind of thing you could probably figure out with tracing enabled if we hook it into the http client 🤔

Probably yeah... I would love to have it in a more user-friendly way in future though, I reckon it'll be very handy.

**Comment by NyaaaWhatsUpDoc** [comment]
> > Honestly this feels like the kind of thing you could probably figure out with tracing enabled if we hook it into the http client 🤔
> 
> Probably yeah... I would love to have it in a more user-friendly way in future though, I reckon it'll be very handy.

the only thing that concerns me is doing this without storing a tonne more data for each item's federation status :thinking:

**Comment by tsmethurst** [comment]
Yep! We'll have to think about a way of doing this nicely, without taking up loads of space in the DB.

**Comment by NyaaaWhatsUpDoc** [comment]
though maybe if we just stored a list of domains with the number of 500-type errors they've returned :thinking: and give an easy way to then block those domains (as they're likely down)

**Comment by hikari-no-yume** [comment]
Hang on…

> GoToSocial uses parallel queues for message delivery[…]. When a message fails to be delivered because of errors on the receiving end, the sender that picked up that particular delivery job will retry for about 2 minutes.

Let's suppose there's three instance we need to deliver a message to, and they're all in the queue: A, B and C. If A is first, and it hits an error, will it keep retrying A before moving onto B and C?

If that's the case, I wonder if there'd be an easy win in sending A to the back of the queue at the moment of failure, and trying B instead. Then you don't even need persistent state to speed up deliveries.

**Comment by tsmethurst** [comment]
> Let's suppose there's three instance we need to deliver a message to, and they're all in the queue: A, B and C. If A is first, and it hits an error, will it keep retrying A before moving onto B and C?

Assuming you're only using one sender then yes. Sending a message to A will block one sender until the retry/backoff loop for A is complete. The default scaling for senders is currently 2 * CPU. So 1 CPU will get you 2 senders, 2 CPUs will get you 4 senders, etc.

Changing the sending system around is not a bad idea, but it would require a fair bit of fiddling given that it's fairly tightly integrated with the `go-fed` library, which doesn't give us a lot of flexibility. Worth doing, but probably won't be something we can do immediately since we have a few things on our plate atm. In fact it might be a good idea to put something like a 'sending strategy' in our go-fed fork...

**Comment by mirabilos** [comment]
ah so that’s why setting it to just one sender to conserve RAM was… not a good idea…

**Comment by daenney** [comment]
We've migrated to Codeberg. Your issue has been imported as https://codeberg.org/superseriousbusiness/gotosocial/issues/2295.
